### ISSUE TITLE
About the Baselines

### ISSUE THREAD

**Issue by xingzhinie** [srs20/STCGAT#1]
Hello, I am very interested in your work，But I have a question. Do you reproduce the baselines mentioned in the article?  In other words, can you make sure the baselines are using speed data？Because as far as I know there is a big difference in the prediction accuracy of speed data and traffic data. 
Hope to get your reply.

**Comment by wengwenchao123** [comment]
> Hello, I am very interested in your work，But I have a question. Do you reproduce the baselines mentioned in the article? In other words, can you make sure the baselines are using speed data？Because as far as I know there is a big difference in the prediction accuracy of speed data and traffic data. Hope to get your reply.

Hello, my friend, have you done any ablation experiments？
I did some ablation experiments, and I found this code very strange
For example, I tried to remove all but the Bidirectional LSTM module and found it was extremely exaggerated, with Mae going 4.3 on PEMS04.

However, the effect is very unstable in the training process. Theoretically, mae fluctuation from model training to later stage should not be large, but I often encounter such situations as MAE adjustment from 4.3 to 6.7, or 8.4 to 5.6. This should not occur in a normal model. The training error model in the later training period should be within 5% at most, and there should not be a huge fluctuation of 30%-40%.

In addition, I made an attempt. Scinet also uses these two datasets, so I tried to remove all scinet models and use a bidirectional LSTM and linear layer, which are consistent with the code I used in this code. The result is that the training effect is very poor, the final MAE is about 30. Although the loss function used by scinet is L1loss, I also tried to use L1loss on stcgat and found that it can run to very low MAE, About 5. In addition, the numerical variation of MAPE is also very strange. Later in training, I have seen 4.64% of the MAPE in step 7 and 10% of the other time steps. Even, I've encountered situations where MAPE is negative.

This indicates that there is a hidden problem in the stcgat code, which leads to the exaggerated effect of the model. But I haven't found the problem yet.

In my experiment, bidirectional LSTM is the main reason for the effect of STCGAT, but the same module with scinet training code has such a big difference. This makes me really hard to believe that STCGAT is OK.

On the other hand, if the bidirectional LSTM is really so exaggerated, it won't be discovered for so many years. Oh, by the way, bidirectional LSTM is problematic in STTCGAT. It should be put into bidirectional LSTM in the form of [batch, step, node] in general, so it is calculated by step. However, in STGCT, his data form is [batch, node, step], which he put into node order, which does not correspond to what he said in his paper.

In short, I think this effect needs to be studied together. After all, from the framework, there is nothing innovative in the model, the only bidirectional LSTM that makes the model outstanding does not play the same role in other models'training mode. Anyway, this paper has a lot of questions.

I think the author may also have found the problem, because the author gave the code link of GitHub in the paper, and now it has become invalid.

**Comment by wengwenchao123** [comment]
> > Hello, I am very interested in your work，But I have a question. Do you reproduce the baselines mentioned in the article? In other words, can you make sure the baselines are using speed data？Because as far as I know there is a big difference in the prediction accuracy of speed data and traffic data. Hope to get your reply.
> 
> Hello, my friend, have you done any ablation experiments？ I did some ablation experiments, and I found this code very strange For example, I tried to remove all but the Bidirectional LSTM module and found it was extremely exaggerated, with Mae going 4.3 on PEMS04.
> 
> However, the effect is very unstable in the training process. Theoretically, mae fluctuation from model training to later stage should not be large, but I often encounter such situations as MAE adjustment from 4.3 to 6.7, or 8.4 to 5.6. This should not occur in a normal model. The training error model in the later training period should be within 5% at most, and there should not be a huge fluctuation of 30%-40%.
> 
> In addition, I made an attempt. Scinet also uses these two datasets, so I tried to remove all scinet models and use a bidirectional LSTM and linear layer, which are consistent with the code I used in this code. The result is that the training effect is very poor, the final MAE is about 30. Although the loss function used by scinet is L1loss, I also tried to use L1loss on stcgat and found that it can run to very low MAE, About 5. In addition, the numerical variation of MAPE is also very strange. Later in training, I have seen 4.64% of the MAPE in step 7 and 10% of the other time steps. Even, I've encountered situations where MAPE is negative.
> 
> This indicates that there is a hidden problem in the stcgat code, which leads to the exaggerated effect of the model. But I haven't found the problem yet.
> 
> In my experiment, bidirectional LSTM is the main reason for the effect of STCGAT, but the same module with scinet training code has such a big difference. This makes me really hard to believe that STCGAT is OK.
> 
> On the other hand, if the bidirectional LSTM is really so exaggerated, it won't be discovered for so many years. Oh, by the way, bidirectional LSTM is problematic in STTCGAT. It should be put into bidirectional LSTM in the form of [batch, step, node] in general, so it is calculated by step. However, in STGCT, his data form is [batch, node, step], which he put into node order, which does not correspond to what he said in his paper.
> 
> In short, I think this effect needs to be studied together. After all, from the framework, there is nothing innovative in the model, the only bidirectional LSTM that makes the model outstanding does not play the same role in other models'training mode. Anyway, this paper has a lot of questions.
> 
> I think the author may also have found the problem, because the author gave the code link of GitHub in the paper, and now it has become invalid.

Also, if you modify the following train code like this

        if epoch % 10 == 0 and epoch != 0:
            mae, rmse, mape = res(model, val_loader, graph, flow_norm)
            print('Average Horizon, MAE:{:.2f}, RMSE: {:.2f}, MAPE: {:.2f}%'.format(mae, rmse, mape))
to
            mae, rmse, mape = res(model, val_loader, graph, flow_norm)
            mae, rmse, mape = res(model, val_loader, graph, flow_norm)
            print('Average Horizon, MAE:{:.2f}, RMSE: {:.2f}, MAPE: {:.2f}%'.format(mae, rmse, mape))

You will find that the two Val results that are the same in theory are not equal in the actual operation, and there are errors.
